crates/proto/src/serialize/binary/encoder.rs

```diff
@@ -25,61 +25,71 @@ mod private {
     use crate::error::{ProtoErrorKind, ProtoResult};
 
     /// A wrapper for a buffer that guarantees writes never exceed a defined set of bytes
-    pub(crate) struct MaximalBuf<'a> {
+    pub(super) struct MaximalBuf<'a> {
         max_size: usize,
         buffer: &'a mut Vec<u8>,
     }
 
     impl<'a> MaximalBuf<'a> {
-        pub(crate) fn new(max_size: u16, buffer: &'a mut Vec<u8>) -> Self {
+        pub(super) fn new(max_size: u16, buffer: &'a mut Vec<u8>) -> Self {
             MaximalBuf {
                 max_size: max_size as usize,
                 buffer,
             }
         }
 
         /// Sets the maximum size to enforce
-        pub(crate) fn set_max_size(&mut self, max: u16) {
+        pub(super) fn set_max_size(&mut self, max: u16) {
             self.max_size = max as usize;
         }
 
-        /// returns an error if the maximum buffer size would be exceeded with the addition number of elements
-        ///
-        /// and reserves the additional space in the buffer
-        pub(crate) fn enforced_write<F>(&mut self, additional: usize, writer: F) -> ProtoResult<()>
-        where
-            F: FnOnce(&mut Vec<u8>),
-        {
-            let expected_len = self.buffer.len() + additional;
+        pub(super) fn write(&mut self, offset: usize, data: &[u8]) -> ProtoResult<()> {
+            debug_assert!(offset <= self.buffer.len());
```

Comment:
what do you think of deduping some of this code by calling `self.reserve()`?

Reply:
I honestly think we're better off not doing that, if we call `reserve()` from `write()` we should add an `offset == self.buffer.len()` in `reserve()` to avoid a double write (resizing with 0, then overwriting with slice bytes). I think these interfaces are just different, and the code is simpler if we keep them separate.

Reply:
👍